platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/MapRenderer.java

```diff
@@ -121,12 +122,38 @@ private native void nativeInitialize(MapRenderer self,
   private void updateFps() {
     frames++;
     long currentTime = System.nanoTime();
-    double fps = 0;
-    if (currentTime - timeElapsed >= 1) {
-      fps = frames / ((currentTime - timeElapsed) / 1E9);
+    long timeCost = currentTime - timeElapsed;
+
+    if (timeCost >= 1) {
+      double fps = frames / ((timeCost) / 1E9);
+      if (fps > preferredFramesPerSecond) {
+        try {
+          long millis = (long) (timeCost / 1E6 * (fps - preferredFramesPerSecond) / fps);
+          Thread.sleep(millis);
+        } catch (InterruptedException ex) {
+          ex.printStackTrace();
+        }
+        long newCurrentTime = System.nanoTime();
+        timeCost = newCurrentTime - timeElapsed;
+        fps = frames / ((timeCost) / 1E9);
+      }
       onFpsChangedListener.onFpsChanged(fps);
       timeElapsed = currentTime;
       frames = 0;
     }
   }
+
+  /**
+   * The preferred frame rate at which this render is rendered,
+   * but it can't excess the ability of device hardware.
+   *
+   * @param preferredFramesPerSecond Can be set to arbitrary integer values.
+   */
+  public void setPreferredFramesPerSecond(int preferredFramesPerSecond) {
+    if (preferredFramesPerSecond <= 0) {
+      // Not valid, just return
+      return;
+    }
+    this.preferredFramesPerSecond = preferredFramesPerSecond;
```

Comment:
it doesn't look like this assignment is synchronized, is it thread safe? what's potential outcome of this being called at the same time while `updateFps()` is executed?

Reply:
If I am not wrong, there is only one Render at the same time, so it will be no thread safe issue.

> what's potential outcome of this being called at the same time while updateFps() is executed?

The fps will be changed immediately.